---
title: "Plotting single cell data with schex"
author: "Saskia Freytag"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{using_schex}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(ggplot2)
theme_set(theme_classic())
```

Reduced dimension plotting is one of the essential tools for the analysis of 
single cell data. However, as the number of cells/nuclei in these these plots 
increases, the usefulness of these plots decreases. Many cells are plotted
on top of each other obscuring information, even when taking advantage of
transparency settings. This package provides binning strategies of cells/nuclei
into hexagon cells. Plotting summarized information of all cells/nuclei in their 
respective hexagon cells presents information without obstructions. The
package seemlessly works with the two most common object classes for the storage
of single cell data; `SingleCellExperiment` from the 
[SingleCellExperiment](https://bioconductor.org/packages/3.9/bioc/html/SingleCellExperiment.html) 
package and `Seurat` from the [Seurat](https://satijalab.org/seurat/) package.

## Load libraries

```{r setup, message=FALSE}
library(igraph)
library(schex)
library(TENxPBMCData)
library(scater)
library(scran)
library(ggrepel)
```

## Setup single cell data

In order to demonstrate the capabilities of the schex package, I will use the
[TENxPBMCData](http://bioconductor.org/packages/release/data/experiment/html/TENxPBMCData.html) 
package and in particular the `pbmc3k` data stored within. This
[dataset](https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc3k) 
was generated with the 10x Genomics platform from human peripheral blood 
mononuclear cells (PBMC). Note that the `TENxPBMCData` package stores data in
the `SingleCellExperiment` object class. For the `schex` package to work single 
cell data is expected to be in a `SingleCellExperiment` or `Seurat` object.

```{r load}
tenx_pbmc3k <- TENxPBMCData(dataset = "pbmc3k")
```

In the  next few sections, I will perform some simple quality control steps
including filtering and normalization. I will then calculate various dimension 
reductions and cluster the data. These steps do by no means constitute
comprehensive handling of the data. For a more detailed guide the reader is 
referred to the following guides:

* [Luecken MD and Theis FJ; Current best practices in single‐cell RNA‐seq analysis:
  a tutorial; Molecular Systems Biology, 2019](https://www.embopress.org/doi/10.15252/msb.20188746)
* [Lun ATL, McCarthy DJ and Marioni JC; A step-by-step workflow for low-level 
  analysis of single-cell RNA-seq data with Bioconductor; F1000 Research, 2016](https://f1000research.com/articles/5-2122)

### Filtering

I filter cells with high mitochondrial content as well as cells with low
library size or feature count.

```{r filter-cells}
rowData(tenx_pbmc3k)$Mito <- grepl("^MT-", rowData(tenx_pbmc3k)$Symbol_TENx)
tenx_pbmc3k <- calculateQCMetrics(tenx_pbmc3k,
  feature_controls = list(Mt = rowData(tenx_pbmc3k)$Mito)
)
tenx_pbmc3k <- tenx_pbmc3k[, !colData(tenx_pbmc3k)$pct_counts_Mt > 50]

libsize_drop <- isOutlier(tenx_pbmc3k$total_counts,
  nmads = 3,type = "lower", log = TRUE
)
feature_drop <- isOutlier(tenx_pbmc3k$total_features_by_counts,
  nmads = 3, type = "lower", log = TRUE
)
tenx_pbmc3k <- tenx_pbmc3k[, !(libsize_drop | feature_drop)]
```

I filter any genes that have 0 count for all cells.

```{r filter-genes}
rm_ind <- calcAverage(tenx_pbmc3k)<0
tenx_pbmc3k <- tenx_pbmc3k[!rm_ind,]
```

### Normalization

I normalize the data by using a simple library size normalization feature.

```{r norm, message=FALSE, warning=FALSE}
tenx_pbmc3k <- normalize(tenx_pbmc3k)
```

### Dimension reduction

I use both Principal Components Analysis (PCA) and Uniform Manifold 
Approximation and Projection (UMAP) in order to obtain reduced dimension 
representations of the data.

```{r dim-red, message=FALSE, warning=FALSE}
tenx_pbmc3k <- runPCA(tenx_pbmc3k)
tenx_pbmc3k <- runUMAP(tenx_pbmc3k, use_dimred = "PCA", n_neighbors = 50)
```

### Clustering

I will cluster the data with on the PCA representation using Louvain clustering.

```{r cluster}
snn_gr <- buildSNNGraph(tenx_pbmc3k, use.dimred = "PCA", k = 50)
clusters <- cluster_louvain(snn_gr)
tenx_pbmc3k$cluster <- factor(clusters$membership)
```

## Plotting single cell data

At this stage in the workflow we usually would like to plot aspects of our data
in one of the reduced dimension representations. Instead of plotting this in an
ordinary fashion, I will demonstrate how schex can provide a better way of 
plotting this.

#### Calculate hexagon cell representation

First, I will calculate the hexagon cell representation for each cell for
a specified dimension reduction representation. I decide to use `nbins=40` which
specifies that I divide my x range into 40 bins. Note that this might be a 
parameter that you want to play around with depending on the number of cells/
nuclei in your dataset. Generally, for more cells/nuclei, `nbins` should be 
increased.

```{r calc-hexbin}
tenx_pbmc3k <- make_hexbin(tenx_pbmc3k, nbins = 40, 
                           dimension_reduction = "UMAP")
```

#### Plot number of cells/nuclei in each hexagon cell

First I plot how many cells are in each hexagon cell. This should be
relatively even, otherwise change the `nbins` parameter in the previous 
calculation.

```{r plot-density, fig.height=7, fig.width=7}
plot_hexbin_density(tenx_pbmc3k)
```

#### Plot meta data in hexagon cell representation

Next I colour the hexagon cells by some meta information, such as the majority 
of cells cluster membership and the median total count in each hexagon cell.

```{r plot-meta, fig.height=7, fig.width=7}
plot_hexbin_meta(tenx_pbmc3k, col="cluster", action="majority")
plot_hexbin_meta(tenx_pbmc3k, col="total_counts", action="median")
```

While for plotting the cluster membership the outcome is not too different from
the classic plot, it is much easier to observe differences in the total count. 

```{r plot-meta-trad, fig.height=7, fig.width=7}
plotUMAP(tenx_pbmc3k, colour_by="cluster")
plotUMAP(tenx_pbmc3k, colour_by="total_counts")
```

For convenience there is also a function that allows the calculation of label
positions for factor variables. These can be overlayed with the package 
`ggrepel`.

```{r plot-meta-label, message=FALSE, fig.height=7, fig.width=7}
label_df <- make_hexbin_label(tenx_pbmc3k, col="cluster")
pp <- plot_hexbin_meta(tenx_pbmc3k, col="cluster", action="majority") 
pp + ggrepel::geom_label_repel(data = label_df, aes(x=x, y=y, label = label), 
                            colour="black",  label.size = NA, fill = NA)  
```

#### Plot gene expression in hexagon cell representation

Finally, I will visualize the gene expression of the POMGNT1 gene in the 
hexagon cell representation.

```{r plot-gene, fig.height=7, fig.width=7}
gene_id <-"ENSG00000085998"
gene_symbol <- rowData(tenx_pbmc3k)$Symbol_TENx[
  which(rownames(tenx_pbmc3k)==gene_id)]
plot_hexbin_gene(tenx_pbmc3k, type="logcounts", gene=gene_id, 
                 action="mean", xlab="UMAP1", ylab="UMAP2", 
                 title=paste0("Mean of ", gene_symbol))
```

Again it is much easier to observe differences in gene expression using the 
hexagon cell representation than the classic representation.

```{r plot-gene-trad, fig.height=7, fig.width=7}
plotUMAP(tenx_pbmc3k, by_exprs_values="logcounts", colour_by=gene_id)
```
